src/python/WMCore/MicroService/Unified/MSRuleCleaner.py

```diff
@@ -528,18 +568,60 @@ def getMSOutputTransferInfo(self, wflow):
+        """
+        Used to resolve parent datasets for a workflow.
+        :param  wflow:   A MSRuleCleaner workflow representation
+        :return:         The workflow object
+        """
+        if wflow['InputDataset'] and wflow['IncludeParents']:
+            childDataset = wflow['InputDataset']
+            parentDataset = findParent([childDataset], self.msConfig['dbsUrl'])
+            if parentDataset:
+                wflow['ParentDataset'].append(parentDataset[childDataset])
+                wflow['ParentDataset'] = list(set(wflow['ParentDataset']))
+                msg = "Found parent %s for input dataset %s for workflow: %s "
+                self.logger.debug(msg, parentDataset, wflow['InputDataset'], wflow['RequestName'])
+            else:
+                msg = "Could not find parent for input dataset: %s for workflows: %s"
+                self.logger.error(msg, wflow['InputDataset'], wflow['RequestName'])
+        return wflow
+
+    def getRucioRules(self, wflow, dataSpec, gran, rucioAcct):
         """
         Queries Rucio and builds the relevant list of blocklevel rules for
         the given workflow
         :param  wflow:   A MSRuleCleaner workflow representation
         :param  gran:    Data granularity to search for Rucio rules. Possible values:
-                        'block' || 'container'
+                         'block' or 'container'
+        :param dataSpec: Data specification. Could be one of:
+                         [primary, secondary, parent, output]
         :return:         The workflow object
         """
         currPline = wflow['PlineMarkers'][-1]
-        # Find all the output placement rules created by the agents
-        for dataCont in wflow['OutputDatasets']:
+
+        # Determine the proper container list we are about to work with:
+        contList = None
+        if dataSpec == 'primary':
+            contList = [wflow['InputDataset']]
+            checkGlobalLocks = True
+        elif dataSpec == 'secondary':
+            contList = list(set(wflow['MCPileup']) | set(wflow['DataPileup']))
+            checkGlobalLocks = True
+        elif dataSpec == 'parent':
+            contList = wflow['ParentDataset']
+            checkGlobalLocks = True
+        elif dataSpec == 'output':
+            contList = wflow['OutputDatasets']
+            checkGlobalLocks = False
+
+        # Find all the data placement rules created by the components:
+        for dataCont in contList:
+            if checkGlobalLocks and dataCont in self.globalLocks:
+                msg = "Found %s dataset: %s in GlobalLocks. NOT considering it for filling the "
+                msg += "RulesToClean list for both container and block level Rules for workflow: %s!"
+                self.logger.info(msg, dataSpec, dataCont, wflow['RequestName'])
+                continue
             if gran == 'container':
                 for rule in self.rucio.listDataRules(dataCont, account=rucioAcct):
```

Comment:
Here I think it would be useful to know the DID and the amount of rules found to be deleted. Can you please add an info log level to this?

Same comment for the `block` granularity below, but we can log just a summary, e.g.:
```
self.logger.info("Found %d block-level rules to be deleted for container %s", blah blah)
```

Reply:
Just to mention, but we used to have that  information. Now, we don't. Remember [1] ?

[1]
https://github.com/dmwm/WMCore/pull/10079

Reply:
Sorry I should have said: "We used to log that  information". It is still in the rule fetched from Rucio, though, but anyway.

Reply:
You have the log message now.
Done.